docs/custom_node_development.md

```diff
@@ -0,0 +1,126 @@
+the metadata of the custom node interfaces: inputs and outputs.
+
+### "CustomNodeParam" struct
+
+Struct CustomNodeParam stores a list of pairs with pointers to parameter and and value strings.
+Each parameter can be references in such object using an index.
+
+### "execute" function
+```
+int execute(const struct CustomNodeTensor* inputs, int inputsCount, struct CustomNodeTensor** outputs, int* outputsCount, const struct CustomNodeParam* params, int paramsCount);
+```
+
+This function implements the data transformation of the custom node. The input data for the function are passed in the form of 
+a pointer to CustomNodeTensor struct object. It includes all the data and pointers to buffers for all custom node inputs.
+The parameter inputsCount pass info about the number of inputs passed that way.
+
+Note that the execute function should not modify the buffers storing the input data because that would alter the data
+which potentially might be used in other pipeline nodes.
+
+The behaviour of custom node execute function can be dependent on the node parameters set in OVMS configuration.
+They are passed to execute function in `params` argument. paramsCount pass info about the number of parameters configured.
+
+The results of the data transformation should be returned by the outputs pointer to a pointer which stores the address of 
+CustomNodeTensor struct. Number of outputs is to be defined during the function execution in the `outputsCount` argument.
+
+Note that during the function execution all the output data buffers needs to be allocated. They will be released by OVMS after 
+the request processing is completed and returned to the user. The cleanup is triggered by calling `release` function.
+
+Execute function returns an integer which value defines the success (0) or failure (1). When the function 
+reports error, the pipeline execution is stopped and error is returned to the user.
+A special response is returned (2) when the execute doesn't return any results. It could be the case when object detection 
+model doesn't detect any valid objects in the input image.
+
+### "getInputsInfo" function
+This function returns information about the metadata of the expected inputs. Returned CustomNodeTensorInfo object is used 
+to create a response for getModelMetadata calls. It is also used in the user request validation and the pipeline 
+configuration validation.
+
+Custom nodes can generate the results which have dynamic size depending on the input data and the custom node parameters.
+In such case function getInputsInfo should return value `-1` on the dimension with dynamic size. It could be an input with
+variable resolution or a batch size.
+
+### "getOutputInfo" function
+Similar to previous function but defining the outputs metadata.
+
+### "release" function
+This function is called by OVMS at the end of the pipeline processing. It should clear all memory allocations used during the 
```

Comment:
This function should only call _free_. OVMS decides when to free and what to free.